WooCommerce/Classes/Tools/InfiniteScroll/AsyncPaginationTracker.swift

```diff
@@ -0,0 +1,132 @@
+
+    /// State of loading the next page in `ensureNextPageIsSynced`.
+    enum NextPageSyncState {
+        case syncing
+        case synced
+        case noNextPage
+    }
+
+    /// Default pagination settings.
+    enum Defaults {
+        static let pageFirstIndex = Store.Default.firstPageNumber
+    }
+
+    /// The index of the first page in the API. So far, both Woo and WP.com API have the first page index at 1.
+    private let pageFirstIndex: Int
+
+    /// Indexes of the pages that have been successfully synced.
+    private var pagesSynced = IndexSet()
+
+    /// Indexes of the pages being currently synced.
+    private var pagesBeingSynced = IndexSet()
+
+    /// Whether there might be more pages to fetch from the API, set by the sync function.
+    private var hasNextPage: Bool = true
+
+    /// Returns the highest page number that has been successfully synced, if any.
+    private var highestPageSynced: Int? {
+        pagesSynced.max()
+    }
+
+    /// Returns the highest page number that is currently being synced, if any.
+    private var highestPageBeingSynced: Int? {
+        pagesBeingSynced.max()
+    }
+
+    /// Designated Initializer
+    init(pageFirstIndex: Int = Defaults.pageFirstIndex) {
+        self.pageFirstIndex = pageFirstIndex
+    }
+
+    /// Should be called whenever a scroll position is approaching the end of the list for infinite scroll support.
+    /// This method will:
+    ///     1.  Proceed only if there is next page to sync.
+    ///     2.  Verify if the next page isn't currently being synced.
+    ///     3.  Proceed syncing the next page.
+    func ensureNextPageIsSynced(syncFunction: @escaping SyncFunction) async throws -> NextPageSyncState {
+        guard hasNextPage else {
+            return .noNextPage
+        }
+
+        let nextPage = (highestPageSynced ?? pageFirstIndex - 1) + 1
+        guard !isPageBeingSynced(pageNumber: nextPage) else {
+            return .syncing
+        }
+        do {
+            try await sync(pageNumber: nextPage, syncFunction: syncFunction)
+            return .synced
+        } catch {
+            throw error
+        }
+    }
+
+    /// Resets internal states and resyncs the first page of results.
+    ///
+    func resync(syncFunction: @escaping SyncFunction) async throws {
+        resetInternalState()
+        try await syncFirstPage( syncFunction: syncFunction)
+    }
+
+    /// Syncs the first page of results.
+    ///
+    func syncFirstPage(syncFunction: @escaping SyncFunction) async throws {
+        try await sync(pageNumber: pageFirstIndex, syncFunction: syncFunction)
+    }
+}
+
+// MARK: - Syncing Core
+//
+private extension AsyncPaginationTracker {
+    /// Syncs a given page number.
+    func sync(pageNumber: Int, syncFunction: @escaping SyncFunction) async throws {
+        markAsBeingSynced(pageNumber: pageNumber)
+
+        do {
+            defer {
+                unmarkAsBeingSynced(pageNumber: pageNumber)
+            }
```

Comment:
I had trouble with defer working as I expected when I used it like this in the ItemsListController – are you sure it gets called if we throw an error? Should it be?

Reply:
Yes, if this defer statement is commented out, `loadNextItems_after_itemProvider_throws_error_then_the_same_page_is_requested_next` fails from the page not being marked as synced. `unmarkAsBeingSynced` was originally called after the do/catch, but when an error is thrown it's never reached. Moving it to `defer` gets called in both cases.

From the [Apple doc](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements/#Defer-Statement):

> The statements within the defer statement are executed no matter how program control is transferred. This means that a defer statement can be used, for example, to perform manual resource management such as closing file descriptors, and to perform actions that need to happen even if an error is thrown.

Reply:
Great, thanks for checking. Goodness knows what I was doing wrong before then... perhaps I accidentally had it in the wrong scope or something.